### PR TITLE
Functional slice sampling implementation, and address UGRID problem #2906

### LINKED ISSUE #2906: Slice + UGRID problem
Import the UGRID data in /glade/p/cisl/vast/vapor/data/Source/UGRID/NOAA-geoflow/small

Create a slice with the following rotation parameters:
X = -19.8
Y = 0
Z = 16.2

The slice will not be sampled properly and will stretch outside of the domain.  Note that the Z stretch factor for this dataset is ~4 million.

### PR DESCRIPTION
**Pull request by sgpearse**
Fixes the following issues from the (now closed) PR #2932:

1) SliceRenderer now uses the SliceGridAlongPlane() function, which is located in the VDC library.
2) Addresses the UGRID problem reported in #2906

This is a moderate refactoring of the last implementation and it uses a new set of algorithms for generating convex hulls, and finding the minimal area rectangles of those convex hulls.  These algorithms are implemented in header files located in VAPOR/include/vapor/GTE.

Note: there is a slight problem that still persists with the UGRID dataset.  This is due to the extreme difference in scale between the X/Y axes (which span tens of millions of meters), compared to the Z axis (which spans 1 meter).  This difference in scale introduces floating point error when projecting certain planes between 2D and 3D coordinates.  The data type for spatial coordinates has been changed from float to double in this PR, and this helps the problem.  But there is still slight error in UGRID's extreme case.  See the image below, showing the texture being drawn slightly outside of the domain.
 
If the user changes their Z coordinates to something reasonable, this error goes away.  I've placed a grid file that demonstrates this in /glade/p/cisl/vast/vapor/data/Source/UGRID/NOAA-geoflow/small, called **gridStretchedZ.nc**.  This file has Z coordinates spanning from 0 to 10,000 instead of 1 to 2.  Load the UGRID data with this file instead of grid.nc and error from the 2D/3D projection arithmetic goes away.  

I'm not aware of glm supporting precision higher than double, and I think it's reasonable to require UGRID users to have a more reasonable representation of the Z coordinates.

Screen Shot 2021-12-17 at 2 39 29 PM

Fixes #2906

### CODE REVIEW

File: include/vapor/SliceGridAlongPlane.h
Comment (StasJ) [inline include/vapor/SliceGridAlongPlane.h:7]: I don't know if the VDC library is technically linking against OpenGL.

File: include/vapor/SliceGridAlongPlane.h
Comment (StasJ) [inline include/vapor/SliceGridAlongPlane.h:25]: These probably should not be in the global namespace.

File: include/vapor/SliceGridAlongPlane.h
Comment (StasJ) [inline include/vapor/SliceGridAlongPlane.h:64]: Missing documentation.

**Review by StasJ — CHANGES_REQUESTED** [review]
This PR adds thousands of lines of code from an external library. Shouldn't this be in our third party libs rather than in the Vapor source tree?

File: include/vapor/SliceRenderer.h
Comment (clyne) [inline include/vapor/SliceRenderer.h:80]: Some of these private member variables and functions don't appear to be used. Please delete or ifdef out if needed for diagnostics.

File: lib/render/SliceRenderer.cpp
Comment (clyne) [inline lib/render/SliceRenderer.cpp:174]: It doesn't look like this memory is ever being freed. is there a reason not to use a smart pointer?

File: lib/render/SliceRenderer.cpp
Comment (clyne) [inline lib/render/SliceRenderer.cpp:185]: Good place to use a smart pointer

File: lib/vdc/SliceGridAlongPlane.cpp
Comment (clyne) [inline lib/vdc/SliceGridAlongPlane.cpp:189]: grid3d should be const shouldn't it?

File: lib/vdc/SliceGridAlongPlane.cpp
Comment (clyne) [inline lib/vdc/SliceGridAlongPlane.cpp:194]: remove commented out code

File: include/vapor/SliceGridAlongPlane.h
Comment (clyne) [reply include/vapor/SliceGridAlongPlane.h:25]: Agreed. @sgpearse you don't need to include these definitions in the public header file if they are only used internally by SliceGridAlongPlane.cpp

**Review by clyne — COMMENTED** [review]
Looks pretty good, @sgpearse. Just a few nits. 

Agree with @stasj: is there a reason to keep the Eberly library in the source? Why not drop it into the 3rd party libs?

**Comment by sgpearse** [comment]
@clyne and @StasJ - comments have been addressed.

Regarding the Geometric Tools files - I agree these should be in our third-party libs.  I submitted this code because it was faster and easier.  Updating our libraries takes a considerable amount of time, and it's harder now than it has been in the past.  Resolving #2922 will help.

Either way I'm updating them now.  I'm hopeful this will be smoother than what happened with the Ospray upgrade.  I'll change this PR into a draft until our builds pass platform tests.

File: lib/vdc/SliceGridAlongPlane.cpp
Comment (sgpearse) [reply lib/vdc/SliceGridAlongPlane.cpp]: Fixed.

File: lib/vdc/SliceGridAlongPlane.cpp
Comment (sgpearse) [reply lib/vdc/SliceGridAlongPlane.cpp]: Yep.  Fixed.

File: lib/render/SliceRenderer.cpp
Comment (sgpearse) [reply lib/render/SliceRenderer.cpp]: Fixed.

File: lib/render/SliceRenderer.cpp
Comment (sgpearse) [reply lib/render/SliceRenderer.cpp]: An oversight.  The function prototype I was given in the last review was using a raw pointer and using a smart ptr didn't occur to me.  Fixed.

File: include/vapor/SliceRenderer.h
Comment (sgpearse) [reply include/vapor/SliceRenderer.h:84]: Fixed.

File: include/vapor/SliceGridAlongPlane.h
Comment (sgpearse) [reply include/vapor/SliceGridAlongPlane.h:29]: Fixed.

File: include/vapor/SliceGridAlongPlane.h
Comment (sgpearse) [reply include/vapor/SliceGridAlongPlane.h]: Fixed.

File: include/vapor/SliceGridAlongPlane.h
Comment (sgpearse) [reply include/vapor/SliceGridAlongPlane.h]: Fixed.

File: include/vapor/SliceRenderer.h
Comment (clyne) [reply include/vapor/SliceRenderer.h:84]: There are still unused private member variables, e.g. _newWaySeconds

File: include/vapor/SliceGridAlongPlane.h
Comment (StasJ) [inline include/vapor/SliceGridAlongPlane.h:34]: Wouldn't the grid contain this data?

Comment (sgpearse) [reply include/vapor/SliceGridAlongPlane.h:34]: No, it actually does not.  The Grid class does contain the extents of the data, but this is not have enough information to define the vertices of the texture's triangles.  Many slices can be defined within the extents of a Grid.

File: include/vapor/SliceRenderer.h
Comment (sgpearse) [reply include/vapor/SliceRenderer.h]: My bad.  I took another pass and removed all unused private members and functions.

**Review by clyne — APPROVED** [review]

File: include/vapor/SliceGridAlongPlane.h
Comment (StasJ) [reply include/vapor/SliceGridAlongPlane.h:34]: There may be something I’m missing here, but would that not contradict the TwoDDataRenderer generating a textured triangle mesh from a Grid class?

Comment (sgpearse) [reply include/vapor/SliceGridAlongPlane.h:34]: I'm not completely familiar with how the TwoD renderers work, but I do know that currently they can be only oriented on the XY plane.  I believe this restriction is what allows them to render without explicitly stating their texture vertex locations.

Comment (clyne) [reply include/vapor/SliceGridAlongPlane.h:34]: Just adding a note that we're going to approve this as is (@StasJ need your approval) and if there is an opportunity to refactor to address the issue raised by Stas it can be done as part of #2728

**Review by StasJ — APPROVED** [review]

**Comment by clyne** [comment]
@sgpearse there were a substantial number of commits made to this PR *after* it was approved. Most of them look like DevOps changes. Can you explain?

**Comment by clyne** [comment]
@sgpearse please split into separate PRS: one for code and one for devops

**Comment by sgpearse** [comment]
Fixes #2906

**Comment by sgpearse** [comment]
@clyne @stasj - The DevOps changes have been backed out, and it looks like they've been merged with the acceptance of PR #2958.

Subsequent changes were needed on this branch regarding clang-format, and some new data types that were introduced into main's VAPoR namespace.  Tests are now passing and this PR is approximately where it was two weeks ago.

File: include/vapor/SliceRenderer.h
Comment (clyne) [inline include/vapor/SliceRenderer.h:17]: Why is it necessary to disable clang-format in this relatively cut-n-dry header file ? We've not yet decided what, if anything, we plan to do with clang-format moving forward. We decided to discuss this issue at our post mortem after the release.

File: include/vapor/SliceRenderer.h
Comment (clyne) [inline include/vapor/SliceRenderer.h:53]: remove commented out code.

File: lib/render/SliceRenderer.cpp
Comment (clyne) [inline lib/render/SliceRenderer.cpp:24]: ditto disabling clang-format here.

File: include/vapor/SliceRenderer.h
Comment (sgpearse) [reply include/vapor/SliceRenderer.h]: Removed.  

To answer your question: 
At our meeting two weeks ago we were discussing clang-format.  I said I'd like to disable it for my work in this PR because it was impeding my work.  Nobody objected to it, and it's used on a subjective basis elsewhere in the repo.

File: include/vapor/SliceRenderer.h
Comment (sgpearse) [reply include/vapor/SliceRenderer.h]: Fixed.

File: lib/render/SliceRenderer.cpp
Comment (sgpearse) [reply lib/render/SliceRenderer.cpp]: Removed.

File: include/vapor/SliceRenderer.h
Comment (clyne) [reply include/vapor/SliceRenderer.h]: I remember the conversation, but maybe I misunderstood. I thought there was some particular, isolated problem that you were running into, not planning on wholesale disabling the formatter.

**Review by clyne — APPROVED** [review]

**Pull request merged**
